src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1057,7 +1057,14 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
               case (Apply(Select(receiver, _), _), SingleType(_, sym)) => sym == receiver.symbol
               case _ => false
             }
-            if (!isThisTypeResult) context.warning(tree.pos, "discarded non-Unit value")
+            def wasTypedExpectingUnit = {
+              tree.attachments.contains[TypedExpectingUnitAttachment.type] && {
+                tree.attachments.remove[TypedExpectingUnitAttachment.type] // cleanup
```

Comment:
This doesn't actually clean up, because [this is not mutation](https://github.com/scala/scala/blob/2.13.x/src/reflect/scala/reflect/macros/Attachments.scala#L67). Normally, `tree.getAndRemoveAttachment`, but here it doesn't work to remove the attachment, as the tree will be seen again after adaptation and it will be checked again for expr in statement position warning.